### PR TITLE
coord,dataflow: separate coordinator and dataflow startup

### PR DESCRIPTION
**Pull request by benesch**
Rather than having the coordinator launch a dataflow server, have the
coordinator take in the client to an already-launched dataflow server.
This paves the way for separate dataflow and coordinator processes.

(Plus a bunch of preparatory commits.)

### Motivation

   * This PR refactors existing code.

### Tips for reviewer

Go commit by commit.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR adds a release note for any
  [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

**Comment by benesch** [comment]
@frankmcsherry only the commits explicitly tagged with `dataflow` are for you, I think!

File: src/dataflow/src/server.rs [hunk 1]
```diff
@@ -297,41 +297,104 @@ pub struct Config {
...
     /// Handle to the persistence runtime. None if disabled.
     pub persist: Option<RuntimeClient>,
-    /// Responses to commands should be sent into this channel.
-    pub feedback_tx: mpsc::UnboundedSender<Response>,
+    /// An optional callback that is presented with both halves of the dataflow
+    /// feedback channel on startup. The callback can swap the channel for
+    /// another in order to intercept messages.
+    ///
+    /// For testing purposes only.
+    pub response_interceptor: Option<ResponseInterceptor>,
```
Comment (frankmcsherry) [inline src/dataflow/src/server.rs:310]: Nit, but this feels a bit like a leaky abstraction that we'll want to fix. I prefer it more than the duplicated coord code, but .. I suspect we'll want to figure out how to shoo this out of the abstraction here eventually. If it is expected to be more durable than that, we should probably chat about what the expectations for it are!

**Review by frankmcsherry — APPROVED** [review]
These changes all look good to me. Only question marks are about the potentially leaky abstraction between coord and dataflow, which seem totally fixable in the future and shouldn't block this.

**Review by antiguru — APPROVED** [review]

File: src/dataflow/src/server.rs (hunk 1 above)
Comment (benesch) [reply src/dataflow/src/server.rs:310]: Yeah, 100%! I figure we'll want `trait DataflowClient` in coord at some point soon, and then we can pass in an `InterceptingDataflowClient` or whatever else `coordtest` wants.

File: src/ore/src/now.rs [hunk 2]
```diff
@@ -9,35 +9,74 @@
...
+    Utc.timestamp(dur.as_secs() as i64, dur.subsec_nanos())
+}
+
 /// A function that returns system or mocked time.
-pub type NowFn = fn() -> EpochMillis;
+// This is a newtype so that it can implement `Debug`, as closures don't
+// implement `Debug` by default. It derefs to a callable so that it is
+// ergonomically equivalent to a closure.
+#[derive(Clone)]
+pub struct NowFn(Arc<dyn Fn() -> EpochMillis + Send + Sync>);
```
Comment (mjibson) [inline src/ore/src/now.rs:37]: Maybe this is clear to someone with more rust experience but: does the use of `dyn` in the `Deref` trait impose an additional lookup that wasn't present in the previous version? Or is there something smart enough to compile away the dyn lookup?

**Review by mjibson — APPROVED** [review]

File: src/ore/src/now.rs (hunk 2 above)
Comment (benesch) [reply src/ore/src/now.rs:37]: I don't actually have a good model for what `dyn Fn` compiles down to. 🤷🏽‍♂️

**Pull request merged**